README
This is the public release of libelf-0.8.12, a free ELF object
file access library. If you have problems with applications
that use libelf and work with the commercial (SVR4, Solaris)
version but not with this one, please contact me.

IMPORTANT NOTE: If you have libelf-0.5.2 installed, you probably
have a file .../include/elf.h that contains the single line
``#include <libelf/elf.h>''. REMOVE THIS FILE BEFORE YOU RUN
configure.

Installation is straightforward - the package is autoconf'ed.  Just do
``cd libelf-0.8.12; ./configure; make; make install''.  Header files
will be installed in .../include/libelf/.  If your system does not
provide its own versions of libelf.h, nlist.h or gelf.h, ``make
install'' will add the missing headers.  If you prefer not to have
these files installed in /usr/include, use ``--disable-compat'' and
add ``-I /usr/include/libelf'' to your CFLAGS when compiling
libelf-based programs.

Note to distribution makers:  You can install libelf in a separate root
hierarchy by using the command ``make instroot=/my/root install''.
You should also use the ``--enable-compat'' configure option in that
case, or run ``make instroot=/my/root install-compat'' manually, to
install all the required header files.

If you are running Linux with libc 5.* as the default C library,
and you plan to use the 64-bit functions, you must either use
``-I.../include/libelf'', or remove /usr/include/libelf.h and use
``--enable-compat'' when running configure. Libc 6.* (aka glibc2)
doesn't have its own <libelf.h>, <nlist.h> or <gelf.h>.

You need an ANSI/ISO C compiler to build libelf. Gcc is optimal.

On some systems (in particular, Solaris and all variants of Linux),
``make'' will try to build a shared library. If you run into problems
on your system, please pass ``--disable-shared'' to configure.
If you build a shared library and want it to be installed as
``libelf-0.8.12.so'' rather than ``libelf.so.0.8.12'', please use
``./configure --enable-gnu-names''. Other files, e.g. ``libelf.so'' and
``libelf.so.0'' are NOT affected.

Another configure option, ``--enable-debug'', adds debugging code to
libelf; if you don't run into problems, you will probably not need it.

When creating an ELF shared library, it is possible to add references
to other shared libraries in the DYNAMIC section of the resulting
file. The make variable DEPSHLIBS contains a list of libraries to add.
It is set to ``-lc'' on Linux systems, and empty otherwise. To
override this setting, use something like ``make DEPSHLIBS="-la -lb"''.
For Linux, `-lc' is included automagically.

NLS is available and enabled by default.  To turn it off, pass the
``--disable-nls'' option to configure.

Libelf can use gettext or catgets for accessing message
catalogs.  If gettext is available AND is part of libc (i.e. not
in a separate library), it will be used. Otherwise, configure
will look for catgets.  If you have gettext in a separate
library and want to use it, you should pass the library's name
to configure, e.g. ``LIBS=-lintl ./configure''. Note that you
MUST link your libelf-based applications with -lintl then,
which is probably not what you want, or change the DEPSHLIBS variable
described above (in case you're building a shared library).

If you have GNU gettext 0.10 installed on your system, and if GNU gettext
runs on top of the catgets interface (rather old Linux systems, using
libc5), configure will refuse to use it and use catgets instead. If you
absolutely want to use GNU gettext, go ahead and rebuild it (which is
IMHO a good idea in general in this case):

	cd .../gettext-0.10
	ac_cv_func_catgets=no ac_cv_func_gettext=no ./configure
	make
	make install

After that, return to the libelf build directory, remove
config.cache, and start over.

*** Large File Support (LFS) applications ***

Some 32-bit systems support files that are larger than the address space
of the architecture.  On these, the `off_t' data type may have 32 or
64 bits, depending on the API you choose.  Since off_t is also part of
the libelf API, in particular the Elf_Data and Elf_Arhdr structures,
an application compiled with large file support will need a version of
libelf that has also been compiled with LFS; otherwise, it won't work
correctly.  Similarly, a program compiled without LFS needs a library
compiled without LFS.

Note that libelf is currently unable to process large files on 32-bit
architectures, whether you compile it for LFS or not, for the simple
reason that the files won't fit into the processes' address space.
Therefore, libelf is compiled without LFS by default.  It can of course
read and write ELF files for 64-bit architectures, but they will be
limited in length on a 32-bit system.

You may compile libelf with large file support by setting CPPFLAGS at
configuration time:

	CPPFLAGS=`getconf LFS_CFLAGS` ./configure

But I really, really recommend you don't, because it breaks binary
compatibility with existing libelf based applications.

*** 64-bit support ***

Starting with libelf-0.7.0, libelf also supports 64-bit ELF files.
This is enabled by default unless your system (or your compiler) does
not support 64-bit integers, or lacks 64-bit declarations in <elf.h>.
If you have problems building with 64-bit support, please do

    ./configure --disable-elf64

for the moment, and contact me. Please note that I haven't tested 64-bit
support much.  There are still some unresolved problems, e.g. IRIX
uses different Elf64_Rel and Elf64_Rela structures (they replaced the
r_info member), and the enumeration values for Elf_Type differ from
the commercial (SVR4) implementation of libelf - they broke binary
compatibility for no good reason, and I'm not willing to follow their
footsteps. The result is that libelf-0.7.* ist upward compatible with
libelf-0.6.4 (as it should be) but INCOMPATIBLE WITH SVR4 LIBELF. If you
have both versions installed, you'd better make sure that you link with
the library that matches the <libelf.h> you're #include'ing.

*** Symbol Versioning ***

Libelf >= 0.8.0 supports the data structures and definitions used for
symbol versioning on Solaris and Linux, in particular, the Elfxx_Verdef,
Elfxx_Verdaux, Elfxx_Verneed, Elfxx_Vernaux and Elfxx_Versym structures
and the SHT_XXX_verdef, SHT_XXX_verneed and SHT_XXX_versym section types
(where `xx' is either `32' or `64', and `XXX' is either `SUNW' or `GNU').
Libelf now translates versioning sections to/from their external
representation properly (earlier versions left them in `raw' format,
with the data type set to ELF_T_BYTE).  This may cause problems on
systems which use the same (OS-specific) section types for different
purposes.  The configure program tries to figure out if your OS uses
versioning; if that check fails, you can use

    ./configure --disable-versioning

to turn off versioning translation support.

*** W32 Support ***

There is now some support for building on W32 systems (requires Microsoft
VC++).  In order to build a W32 DLL, cd into the `lib' subdirectory, edit
build.bat if necessary (it needs the path to your compiler binaries) and
run it.  If you're lucky, libelf.dll and the import/export libraries will
be built.  If not, please drop me a line.

I tested it on XP Pro (SP2), using VC++ 2005 Express Edition.
Apparently, Visual Studio .NET 2003 works fine as well.

Various notes regarding the W32 port:

    - When you open() an ELF file, remember to use the O_BINARY flag.
    - You may have to add /MD to the linker command line.

*** Missing things ***

	* There is no documentation.  You can use the Solaris
	  manpages instead (available at http://docs.sun.com/).
	  The ELF file format is described in several places;
	  among them Suns "Linker and Libraries Guide" and the
	  "System V Application Binary Interface" documents;
	  http://www.caldera.com/developer/devspecs/gabi41.pdf and
	  http://www.caldera.com/developer/gabi/ are probably good
	  starting points. Processor-specific documentation is spread
	  across a number of `Processor Supplement' documents, one
	  for each architecture; you'll have to use a search engine to
	  find them.

	* The COFF file format is not understood. This is so obsolete
	  that it will probably never be implemented.

	* nlist(3) is incomplete; the n_type and n_sclass
	  members of struct nl are set to zero even if type
	  (that is, debug) information is available.

	* Libelf does not translate Solaris' `Move' and `Syminfo'
	  sections. You can read them using elf_getdata(), but you'll
	  only get raw (untranslated) bytes.

Changes since 0.8.11:

	* Due to some unfortunate confusion, the elf_getphnum(),
	  elf_getshnum() and elf_getshstrndx() are not compatible
	  between libelf implementations. Therefore, the developers
	  decided to replace them with new functions: elf_getphdrnum(),
	  elf_getshdrnum() and elf_getshdrstrndx(), which will always
	  return -1 on failure and 0 on success. Code using the old
	  interface should be upgraded to increase portability.

Changes since 0.8.10:

	* Fixed a bug in elf_rawfile().
	* If you use ELF_F_LAYOUT together with ELF_F_LAYOUT_OVERLAP,
	  elf_update() will now tolerate overlapping sections.

Changes since 0.8.9:

	* Ported to QNX Neutrino.
	* Fixed Windows build errors.
	* Parallel (make -j) installation should work now.

	* It's now possible to enable and disable select sanity checks
	  libelf performs. Currently, this affects the "NUL terminated
	  string table entry" check performed in elf_strptr(). By
	  default, the function will return an error if the string
	  requested is not properly terminated - because some
	  applications might dump core otherwise. If you configure
	  libelf with `--disable-sanity-checks', however, the check
	  (and, in the future, probably others as well) is disabled
	  by default. You can still turn it on and off at runtime by
	  setting the LIBELF_SANITY_CHECKS environment variable to
	  an integer value:

	      # disable all sanity checks
	      export LIBELF_SANITY_CHECKS=0

	      # enable all sanity checks
	      export LIBELF_SANITY_CHECKS=-1

	  Each bit of the value corresponds to a particular check,
	  so you could use LIBELF_SANITY_CHECKS=1 to enable only
	  the elf_strptr() check. You may also use a value in hex
	  (0x...) or octal (0...) format.
 
Changes since 0.8.8:

	* Improved translator for symbol versioning sections.
	* The W32 library is now built in the `lib' subdirectory.
	* Windows DLLs should work now.

Changes since 0.8.6:

	* added elf_getphnum().
	* added elf_getshnum().
	* added elf_getshstrndx().
	* added elfx_update_shstrndx().
	* handle interrupted reads/writes more gracefully.
	* added (partial) support for unusual e_[ps]hentsize values.
	* fixed the bugs introduced in 0.8.7.

Changes since 0.8.5:

	* added W32 support.
	* added workaround for alignment errors in archive members.
	* my email address has changed again ;)

Changes since 0.8.4:

	* elf_strptr() should now work more safely with fragmented
	  or badly formatted string tables.

Changes since 0.8.3:

	* Fixed a bug in elf_update() that was introduced in 0.8.3.

Changes since 0.8.2:

	* Should compile on MacOSX now.

	* Can read and write files with more than 65280 sections

	* Tries to handle 64-bit ELF files that use 8-byte hash table
	  entries. In particular, libelf tries to guess the data type in
	  elf_getdata(), and doesn't override sh_entsize in elf_update()
	  any longer. If you want the library to pick the entry size,
	  you must set its value to 0 before you call elf_update().

	* No longer dumps core in elf_update() when a versioning section
	  has no data. Instead, it returns an error message. Note that
	  you're supposed to provide a valid d_buf for any section, unless
	  it's empty or has SHT_NOBITS type.

	* Building a shared library is now the default (if supported).


ChangeLog
Tue Jul  7 19:45:12 CEST 2009, Michael Riepe <libelf@mr511.de>

	* README:
	* VERSION:
		update for 0.8.12 release.

	* lib/libelf.h:
	* lib/x.elfext.c:
		add new and deprecate old functions.
 
Fri May 22 19:17:14 CEST 2009, Michael Riepe <libelf@mr511.de>

	* README:
	* VERSION:
		update for 0.8.11 release.

	* lib/libelf.h:
		add ELF_F_LAYOUT_OVERLAP flag.

	* lib/rawfile.c:
		fix re-read zero size bug.

	* lib/update.c:
		honor ELF_F_LAYOUT_OVERLAP flag.

Fri Sep  7 14:04:20 CEST 2007, Michael Riepe <libelf@mr511.de>

	* acconfig.h:
		add ENABLE_SANITY_CHECKS.

	* aclocal.m4:
		fix --enable-gnu-names.

	* configure.in:
		add --enable-sanity-checks.
		fix --enable-extended-format.

	* lib/data.c:
		add _elf_sanity_checks variable.

	* lib/private.h:
		declare _elf_sanity_checks and constants.

	* lib/strptr.c:
		enable/disable sanity check.

	* lib/version.c:
		set _elf_sanity_checks from $LIBELF_SANITY_CHECKS.

Fri Jun 29 23:27:15 CEST 2007, Michael Riepe <libelf@mr511.de>

	* lib/Makefile.in:
		improved make -jX patch.

Wed Jun 20 08:04:30 CEST 2007, Michael Riepe <libelf@mr511.de>

	* lib/Makefile.in:
		add "make -jX install" patch by Joel Martin.

Tue Nov 21 21:21:12 CET 2006, Michael Riepe <libelf@mr511.de>

	* lib/Makefile.w32:
		fix Windows compilation bug.

Thu Sep  7 17:55:42 CEST 2006, Michael Riepe <libelf@mr511.de>

	* acconfig.h:
	* aclocal.m4:
	* configure.in:
	* lib/config.h.w32:
	* lib/gelf.h:
	* lib/private.h:
	* lib/sys_elf.h.in:
	* lib/sys_elf.h.w32:
		port to QNX Neutrino, thanks to darkelf.

Fri Aug 25 14:46:34 CEST 2006, Michael Riepe <libelf@mr511.de>

	* Makefile.in:
		add trackinstall target.

Mon Aug 21 20:26:47 CEST 2006, Michael Riepe <libelf@mr511.de>

	* Makefile.in:
		drop w32 from DISTSUBDIRS.
	* lib/Makefile.in:
		add new files to DISTFILES.
	* lib/Makefile.w32:
	* lib/build.bat:
	* lib/config.h.w32:
	* lib/libelf.def:
	* lib/sys_elf.h.w32:
		adopted from w32 subdirectory.

Fri Aug 18 02:04:58 CEST 2006, Michael Riepe <libelf@mr511.de>

	* lib/begin.c:
		let getnum return a size_t.
	* lib/libelf.h:
		replace __value because it's a reserved word in VC++ 2005.
	* lib/nlist.c:
		don't declare open() on W32.
	* lib/private.h:
		use <io.h> on W32.
	* w32/Makefile.w32:
		fix W32 DLL build.
	* w32/build.bat:
		add more examples for vcvars32.bat location.

Fri Jul 28 00:56:00 CEST 2006, Michael Riepe <libelf@mr511.de>

	* lib/32.xlatetof.c:
	* lib/64.xlatetof.c:
		check for dsize == -1.
	* lib/verdef.h:
	* lib/verneed.h:
		improve section translators.

Tue Jul 11 18:53:00 CEST 2006, Michael Riepe <libelf@mr511.de>

	* w32/libelf.def:
		add missing functions.

Sat Jul  8 00:50:00 CEST 2006, Michael Riepe <libelf@mr511.de>

	* VERSION:
		bump up to 0.8.9.

Sat Jul  8 00:17:00 CEST 2006, Michael Riepe <libelf@mr511.de>

	* lib/32.newehdr.c:
		make return value compatible with Solaris.
	* lib/32.newphdr.c:
		handle 65535+ segments.
		make return value compatible with Solaris.
	* lib/cook.c:
		handle 65535+ segments.
	* lib/elf_repl.h:
		add new definitions.
	* lib/libelf.h:
		add/rename functions.
	* lib/newscn.c:
		fix section creation (was broken in 0.8.7).
	* lib/private.h:
		add SHN_XINDEX and PN_XNUM in case they're missing.
		centrally define LIBELF_SUCCESS and LIBELF_FAILURE.
	* lib/update.c:
		handle 65535+ segments.
		use elf->e_phnum internally.
	* lib/x.elfext.c:
		add elf_getphnum().
		rename elfx_get_shnum() and elfx_get_shstrndx().
		make return values compatible with Solaris.

Fri Jul  7 19:01:04 CEST 2006, Michael Riepe <libelf@mr511.de>

	* VERSION:
		bump up to 0.8.8.

Fri Jul  7 18:27:25 CEST 2006, Michael Riepe <libelf@mr511.de>

	* lib/Makefile.in:
		add lib/x.elfext.c.
	* lib/libelf.h:
		add functions from lib/x.elfext.c.
	* lib/newscn.c:
		simplify _elf_update_shnum().

Tue Apr 25 16:26:39 CEST 2006, Michael Riepe <libelf@mr511.de>

	* lib/gelf.h:
	* lib/libelf.h:
	* lib/nlist.h:
	* lib/private.h:
		add workaround for broken compilers.

Mon Apr 24 16:24:32 CEST 2006, Michael Riepe <libelf@mr511.de>

	* po/de.po:
		update.

Fri Apr 21 19:17:46 CEST 2006, Michael Riepe <libelf@mr511.de>

	* acconfig.h:
	* configure.in:
		add --enable-extended-format.
	* aclocal.m4:
		search for msgmerge.
	* lib/cook.c:
		change _elf_item buffering.
		handle extended format (with unusual entsize).
	* lib/errors.h:
	 	add ERROR_EHDR_SHENTSIZE and ERROR_EHDR_PHENTSIZE.
	* po/Makefile.in:
		use msgmerge instead of tupdate.
 
Thu Oct 20 21:08:02 CEST 2005, Michael Riepe <libelf@mr511.de>

	* lib/input.c:
	* lib/update.c:
		handle partial reads and writes.

Tue Aug 16 01:48:17 CEST 2005, Michael Riepe <libelf@mr511.de>

	* lib/begin.c:
		add workaround for archive member misalignment.
	* VERSION:
		bump up to 0.8.7

Tue Jul 19 11:56:26 CEST 2005, Michael Riepe <libelf@mr511.de>

	* README:
	* w32/build.bat:
		update.
	* w32/libelf.def:
		fix syntax.

Tue Jun 28 00:31:24 CEST 2005, Michael Riepe <libelf@mr511.de>

	* Makefile.in:
		remove superfluous slash.

Tue Jun 21 03:58:47 CEST 2005, Michael Riepe <libelf@mr511.de>

	* lib/Makefile.in:
		get rid of lib/pic subdir.

Sat May 21 17:39:28 CEST 2005, Michael Riepe <libelf@mr511.de>

	* (global):
		remove my e-mail address from all copyright clauses.

Sun May 15 23:08:30 CEST 2005, Michael Riepe <libelf@mr511.de>

	* configure.in:
		check if $CC can copile <elf.h>.
	* lib/private.h:
		#include <stdint.h> before <sys/types.h> (fixes glibc bug).

Sun May  8 23:40:35 CEST 2005, Michael Riepe <libelf@mr511.de>

	* Makefile.in:
		add instroot variable.
		install libelf.pc.
	* configure.in:
		create libelf.pc.

Sun Mar 20 15:41:22 CET 2005, Michael Riepe <libelf@mr511.de>

	* (global):
		change my e-mail address.

Fri Jan 28 23:09:57 CET 2005, Michael Riepe <michael@stud.uni-hannover.de>

	* po/Makefile.in:
		use modified gmo2msg.
	* po/gmo2msg.c:
		make gmo2msg output more portable.


lib/private.h
/*
 * private.h - private definitions for libelf.
 * Copyright (C) 1995 - 2007 Michael Riepe
 * 
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 * 
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 * 
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA
 */

/* @(#) $Id: private.h,v 1.39 2008/05/23 08:15:35 michael Exp $ */

#ifndef _PRIVATE_H
#define _PRIVATE_H

#define __LIBELF_INTERNAL__ 1

#if HAVE_CONFIG_H
# include <config.h>
#endif /* HAVE_CONFIG_H */

/*
 * Workaround for GLIBC bug:
 * include <stdint.h> before <sys/types.h>
 */
#if HAVE_STDINT_H
#include <stdint.h>
#endif
#include <sys/types.h>

#if STDC_HEADERS
# include <stdlib.h>
# include <string.h>
#else /* STDC_HEADERS */
extern void *malloc(), *realloc();
extern void free(), bcopy(), abort();
extern int strcmp(), strncmp(), memcmp();
extern void *memcpy(), *memmove(), *memset();
#endif /* STDC_HEADERS */

#if defined(_WIN32)
#include <io.h>
#else
#if HAVE_UNISTD_H
# include <unistd.h>
#else /* HAVE_UNISTD_H */
extern int read(), write(), close();
extern off_t lseek();
#if HAVE_FTRUNCATE
extern int ftruncate();
#endif /* HAVE_FTRUNCATE */
#endif /* HAVE_UNISTD_H */
#endif /* defined(_WIN32) */

#ifndef SEEK_SET
#define SEEK_SET	0
#endif /* SEEK_SET */
#ifndef SEEK_CUR
#define SEEK_CUR	1
#endif /* SEEK_CUR */
#ifndef SEEK_END
#define SEEK_END	2
#endif /* SEEK_END */

#if !HAVE_MEMCMP
# define memcmp	strncmp
#endif /* !HAVE_MEMCMP */
#if !HAVE_MEMCPY
# define memcpy(d,s,n)	bcopy(s,d,n)
#endif /* !HAVE_MEMCPY */
#if !HAVE_MEMMOVE
# define memmove(d,s,n)	bcopy(s,d,n)
#endif /* !HAVE_MEMMOVE */

#if !HAVE_MEMSET
# define memset _elf_memset
extern void *_elf_memset();
#endif /* !HAVE_MEMSET */

#if HAVE_STRUCT_NLIST_DECLARATION
# define nlist __override_nlist_declaration
#endif /* HAVE_STRUCT_NLIST_DECLARATION */

#if __LIBELF_NEED_LINK_H
# include <link.h>
#elif __LIBELF_NEED_SYS_LINK_H
# include <sys/link.h>
#endif /* __LIBELF_NEED_LINK_H */

#include <libelf.h>

#if HAVE_STRUCT_NLIST_DECLARATION
# undef nlist
#endif /* HAVE_STRUCT_NLIST_DECLARATION */

#if __LIBELF64
#include <gelf.h>
#endif /* __LIBELF64 */

typedef struct Scn_Data Scn_Data;

/*
 * ELF descriptor
 */
struct Elf {
    /* common */
    size_t	e_size;			/* file/member size */
    size_t	e_dsize;		/* size of memory image */
    Elf_Kind	e_kind;			/* kind of file */
    char*	e_data;			/* file/member data */
    char*	e_rawdata;		/* file/member raw data */
    size_t	e_idlen;		/* identifier size */
    int		e_fd;			/* file descriptor */
    unsigned	e_count;		/* activation count */
    /* archive members (still common) */
    Elf*	e_parent;		/* NULL if not an archive member */
    size_t	e_next;			/* 0 if not an archive member */
    size_t	e_base;			/* 0 if not an archive member */
    Elf*	e_link;			/* next archive member or NULL */
    Elf_Arhdr*	e_arhdr;		/* archive member header or NULL */
    /* archives */
    size_t	e_off;			/* current member offset (for elf_begin) */
    Elf*	e_members;		/* linked list of active archive members */
    char*	e_symtab;		/* archive symbol table */
    size_t	e_symlen;		/* length of archive symbol table */
    char*	e_strtab;		/* archive string table */
    size_t	e_strlen;		/* length of archive string table */
    /* ELF files */
    unsigned	e_class;		/* ELF class */
    unsigned	e_encoding;		/* ELF data encoding */
    unsigned	e_version;		/* ELF version */
    char*	e_ehdr;			/* ELF header */
    char*	e_phdr;			/* ELF program header table */
    size_t	e_phnum;		/* size of program header table */
    Elf_Scn*	e_scn_1;		/* first section */
    Elf_Scn*	e_scn_n;		/* last section */
    unsigned	e_elf_flags;		/* elf flags (ELF_F_*) */
    unsigned	e_ehdr_flags;		/* ehdr flags (ELF_F_*) */
    unsigned	e_phdr_flags;		/* phdr flags (ELF_F_*) */
    /* misc flags */
    unsigned	e_readable : 1;		/* file is readable */
    unsigned	e_writable : 1;		/* file is writable */
    unsigned	e_disabled : 1;		/* e_fd has been disabled */
    unsigned	e_cooked : 1;		/* e_data was modified */
    unsigned	e_free_syms : 1;	/* e_symtab is malloc'ed */
    unsigned	e_unmap_data : 1;	/* e_data is mmap'ed */
    unsigned	e_memory : 1;		/* created by elf_memory() */
    /* magic number for debugging */
    long	e_magic;
};

#define ELF_MAGIC	0x012b649e

#define INIT_ELF	{\
    /* e_size */	0,\
    /* e_dsize */	0,\
    /* e_kind */	ELF_K_NONE,\
    /* e_data */	NULL,\
    /* e_rawdata */	NULL,\
    /* e_idlen */	0,\
    /* e_fd */		-1,\
    /* e_count */	1,\
    /* e_parent */	NULL,\
    /* e_next */	0,\
    /* e_base */	0,\
    /* e_link */	NULL,\
    /* e_arhdr */	NULL,\
    /* e_off */		0,\
    /* e_members */	NULL,\
    /* e_symtab */	NULL,\
    /* e_symlen */	0,\
    /* e_strtab */	NULL,\
    /* e_strlen */	0,\
    /* e_class */	ELFCLASSNONE,\
    /* e_encoding */	ELFDATANONE,\
    /* e_version */	EV_NONE,\
    /* e_ehdr */	NULL,\
    /* e_phdr */	NULL,\
    /* e_phnum */	0,\
    /* e_scn_1 */	NULL,\
    /* e_scn_n */	NULL,\
    /* e_elf_flags */	0,\
    /* e_ehdr_flags */	0,\
    /* e_phdr_flags */	0,\
    /* e_readable */	0,\
    /* e_writable */	0,\
    /* e_disabled */	0,\
    /* e_cooked */	0,\
    /* e_free_syms */	0,\
    /* e_unmap_data */	0,\
    /* e_memory */	0,\
    /* e_magic */	ELF_MAGIC\
}

/*
 * Section descriptor
 */
struct Elf_Scn {
    Elf_Scn*	s_link;			/* pointer to next Elf_Scn */
    Elf*	s_elf;			/* pointer to elf descriptor */
    size_t	s_index;		/* number of this section */
    unsigned	s_scn_flags;		/* section flags (ELF_F_*) */
    unsigned	s_shdr_flags;		/* shdr flags (ELF_F_*) */
    Scn_Data*	s_data_1;		/* first data buffer */
    Scn_Data*	s_data_n;		/* last data buffer */
    Scn_Data*	s_rawdata;		/* raw data buffer */
    /* data copied from shdr */
    unsigned	s_type;			/* section type */
    size_t	s_offset;		/* section offset */
    size_t	s_size;			/* section size */
    /* misc flags */
    unsigned	s_freeme : 1;		/* this Elf_Scn was malloc'ed */
    /* section header */
    union {
#if __LIBELF64
	Elf64_Shdr	u_shdr64;
#endif /* __LIBELF64 */
	Elf32_Shdr	u_shdr32;
    }		s_uhdr;
    /* magic number for debugging */
    long	s_magic;
};
#define s_shdr32	s_uhdr.u_shdr32
#define s_shdr64	s_uhdr.u_shdr64

#define SCN_MAGIC	0x012c747d

#define INIT_SCN	{\
    /* s_link */	NULL,\
    /* s_elf */		NULL,\
    /* s_index */	0,\
    /* s_scn_flags */	0,\
    /* s_shdr_flags */	0,\
    /* s_data_1 */	NULL,\
    /* s_data_n */	NULL,\
    /* s_rawdata */	NULL,\
    /* s_type */	SHT_NULL,\
    /* s_offset */	0,\
    /* s_size */	0,\
    /* s_freeme */	0,\
    /* s_uhdr */	{{0,}},\
    /* s_magic */	SCN_MAGIC\
}

/*
 * Data descriptor
 */
struct Scn_Data {
    Elf_Data	sd_data;		/* must be first! */
    Scn_Data*	sd_link;		/* pointer to next Scn_Data */
    Elf_Scn*	sd_scn;			/* pointer to section */
    char*	sd_memdata;		/* memory image of section */
    unsigned	sd_data_flags;		/* data flags (ELF_F_*) */
    /* misc flags */
    unsigned	sd_freeme : 1;		/* this Scn_Data was malloc'ed */
    unsigned	sd_free_data : 1;	/* sd_memdata is malloc'ed */
    /* magic number for debugging */
    long	sd_magic;
};

#define DATA_MAGIC	0x01072639

#define INIT_DATA	{\
    {\
    /* d_buf */		NULL,\
    /* d_type */	ELF_T_BYTE,\
    /* d_size */	0,\
    /* d_off */		0,\
    /* d_align */	0,\
    /* d_version */	EV_NONE\
    },\
    /* sd_link */	NULL,\
    /* sd_scn */	NULL,\
    /* sd_memdata */	NULL,\
    /* sd_data_flags */	0,\
    /* sd_freeme */	0,\
    /* sd_free_data */	0,\
    /* sd_magic */	DATA_MAGIC\
}

/*
 * Private status variables
 */
extern unsigned _elf_version;
extern int _elf_errno;
extern int _elf_fill;
extern int _elf_sanity_checks;
#define SANITY_CHECK_STRPTR	(1u << 0)

/*
 * Private functions
 */
extern void *_elf_read __P((Elf*, void*, size_t, size_t));
extern void *_elf_mmap __P((Elf*));
extern int _elf_cook __P((Elf*));
extern char *_elf_getehdr __P((Elf*, unsigned));
extern char *_elf_getphdr __P((Elf*, unsigned));
extern Elf_Data *_elf_xlatetom __P((const Elf*, Elf_Data*, const Elf_Data*));
extern Elf_Type _elf_scn_type __P((unsigned));
extern size_t _elf32_xltsize __P((const Elf_Data *__src, unsigned __dv, unsigned __encode, int __tof));
extern size_t _elf64_xltsize __P((const Elf_Data *__src, unsigned __dv, unsigned __encode, int __tof));
extern int _elf_update_shnum(Elf *__elf, size_t __shnum);
extern Elf_Scn *_elf_first_scn(Elf *__elf);

/*
 * Special translators
 */
extern size_t _elf_verdef_32L11_tof __P((unsigned char *dst, const unsigned char *src, size_t n));
extern size_t _elf_verdef_32L11_tom __P((unsigned char *dst, const unsigned char *src, size_t n));
extern size_t _elf_verdef_32M11_tof __P((unsigned char *dst, const unsigned char *src, size_t n));
extern size_t _elf_verdef_32M11_tom __P((unsigned char *dst, const unsigned char *src, size_t n));
extern size_t _elf_verdef_64L11_tof __P((unsigned char *dst, const unsigned char *src, size_t n));
extern size_t _elf_verdef_64L11_tom __P((unsigned char *dst, const unsigned char *src, size_t n));
extern size_t _elf_verdef_64M11_tof __P((unsigned char *dst, const unsigned char *src, size_t n));
extern size_t _elf_verdef_64M11_tom __P((unsigned char *dst, const unsigned char *src, size_t n));
extern size_t _elf_verneed_32L11_tof __P((unsigned char *dst, const unsigned char *src, size_t n));
extern size_t _elf_verneed_32L11_tom __P((unsigned char *dst, const unsigned char *src, size_t n));
extern size_t _elf_verneed_32M11_tof __P((unsigned char *dst, const unsigned char *src, size_t n));
extern size_t _elf_verneed_32M11_tom __P((unsigned char *dst, const unsigned char *src, size_t n));
extern size_t _elf_verneed_64L11_tof __P((unsigned char *dst, const unsigned char *src, size_t n));
extern size_t _elf_verneed_64L11_tom __P((unsigned char *dst, const unsigned char *src, size_t n));
extern size_t _elf_verneed_64M11_tof __P((unsigned char *dst, const unsigned char *src, size_t n));
extern size_t _elf_verneed_64M11_tom __P((unsigned char *dst, const unsigned char *src, size_t n));

/*
 * Private data
 */
extern const Elf_Scn _elf_scn_init;
extern const Scn_Data _elf_data_init;
extern const size_t _elf_fmsize[2][EV_CURRENT - EV_NONE][ELF_T_NUM][2];

/*
 * Access macros for _elf_fmsize[]
 */
#define _fmsize(c,v,t,w)	\
	(_elf_fmsize[(c)-ELFCLASS32][(v)-EV_NONE-1][(t)-ELF_T_BYTE][(w)])
#define _fsize(c,v,t)		_fmsize((c),(v),(t),1)
#define _msize(c,v,t)		_fmsize((c),(v),(t),0)

/*
 * Various checks
 */
#define valid_class(c)		((c) >= ELFCLASS32 && (c) <= ELFCLASS64)
#define valid_encoding(e)	((e) >= ELFDATA2LSB && (e) <= ELFDATA2MSB)
#define valid_version(v)	((v) > EV_NONE && (v) <= EV_CURRENT)
#define valid_type(t)		((unsigned)(t) < ELF_T_NUM)

/*
 * Error codes
 */
enum {
#define __err__(a,b)	a,
#include <errors.h>		/* include constants from errors.h */
#undef __err__
ERROR_NUM
};

#define seterr(err)	(_elf_errno = (err))

/*
 * Sizes of data types (external representation)
 * These definitions should be in <elf.h>, but...
 */
#ifndef ELF32_FSZ_ADDR
# define ELF32_FSZ_ADDR		4
# define ELF32_FSZ_HALF		2
# define ELF32_FSZ_OFF		4
# define ELF32_FSZ_SWORD	4
# define ELF32_FSZ_WORD		4
#endif /* ELF32_FSZ_ADDR */
#ifndef ELF64_FSZ_ADDR
# define ELF64_FSZ_ADDR		8
# define ELF64_FSZ_HALF		2
# define ELF64_FSZ_OFF		8
# define ELF64_FSZ_SWORD	4
# define ELF64_FSZ_SXWORD	8
# define ELF64_FSZ_WORD		4
# define ELF64_FSZ_XWORD	8
#endif /* ELF64_FSZ_ADDR */

/*
 * More missing pieces, in no particular order
 */
#ifndef SHT_SYMTAB_SHNDX
#define SHT_SYMTAB_SHNDX	18
#endif /* SHT_SYMTAB_SHNDX */

#ifndef SHN_XINDEX
#define SHN_XINDEX		0xffff
#endif /* SHN_XINDEX */

#ifndef PN_XNUM
#define PN_XNUM			0xffff
#endif /* PN_XNUM */

/*
 * Debugging
 */
#if ENABLE_DEBUG
extern void __elf_assert __P((const char*, unsigned, const char*));
# if (__STDC__ + 0)
#  define elf_assert(x)	do{if(!(x))__elf_assert(__FILE__,__LINE__,#x);}while(0)
# else /* __STDC__ */
#  define elf_assert(x)	do{if(!(x))__elf_assert(__FILE__,__LINE__,"x");}while(0)
# endif /* __STDC__ */
#else /* ENABLE_DEBUG */
# define elf_assert(x)	do{}while(0)
#endif /* ENABLE_DEBUG */

/*
 * Return values for certain functions
 */
#define LIBELF_SUCCESS	1
#define LIBELF_FAILURE	0

#endif /* _PRIVATE_H */
